박스오피스 [STEP 1] BMO,Serena

serena0720 opened the pull request:
## 팀원 : Serena🐷, BMO🤖

@GREENOVER 
안녕하세요! 그린🌳

박스오피스 프로젝트 Step1 PR 보냅니다.
2주동안 리뷰 잘 부탁드립니다~ 

-----

## 구현한 내용
`box_office_sample.json` 파일을 사용하여 모델 생성 및 `Decode`를 하였습니다.
> 주요 내용 : `JSONDecoder`, `Decodable`, `UnitTest`

-----


## 고민한 부분
### 🔥 CodingKey
- 초반 코드 작성 시 `decode` 과정에서 알 수 없는 에러가 계속 발생하였습니다. `decode` 과정에서 어느 부분에서 잘못되었는지를 찾기 위해 코드를 처음부터 다시 작성하다보니, `CodingKey`를 잘못 작성하여 `JSON decode`가 되지 않았다는 것을 발견했습니다.
- `CodingKey`를 사용할 때 프로퍼티명을 `swift`에서 사용할 이름으로 지정하고, `enum case`의 값으로 기존의 `JSON` 키 값의 이름을 지정해야합니다. 하지만 이를 반대로 작성하였더니, 알 수 없는 에러가 발생하여 이를 디버깅하는데 많은 시간을 소요하였습니다.
> CodingKey 잘못 작성 예시
```swift
struct BoxOfficeResult: Decodable {
    let boxofficeType: String
    ...
    
    enum CodingKeys: String, CodingKey {
        case boxofficeType = "boxOfficeType"
        ...
    }
}
```

> CodingKey 올바른 예시
```swift
struct BoxOfficeResult: Decodable {
    let boxOfficeType: String
    ...
    
    private enum CodingKeys: String, CodingKey {
        case boxOfficeType = "boxofficeType"
        ...
    }
}
```

<br>

### 🔥 XCTFail
- 테스트에서 파일 디코딩 중 실패했을 때 테스트 메서드를 `retrun`하면 테스트가 `Success`처리 되는것을 확인했습니다.
- 테스트가 꼭 `Then`과정까지 도달하지 않더라도 `Given`과 `When`과정 또한 테스트에 적절하지 않은 값이 들어오거나, 값이 처리되는 과정에 대한 처리가 필요하다고 생각했습니다.
    ```swift
    func test_box_office_sample_json_파일을_디코딩_할_수_있다() {
        // Given
        guard let result: BoxOffice = JSONDecoder.decode(fileName: "box_office_sample") else {
            XCTFail("파일명 'box_office_sample'로 JSON 디코딩 할 수 없습니다.")
            return
        }

        // When

        // Then
        XCTAssertNotNil(result)
    }
    ```
- `XCTFail` 메서드를 찾아 테스트 진행 중 적절하지 않은 부분에 삽입해 주었습니다. `XCTFail`의 설명은 다음과 같습니다.
    > This function generates a failure immediately and unconditionally.
    > (이 함수는 즉시 무조건 실패를 생성합니다.)


-----

## 참고자료
- [🍎 Developer Apple - XCTFail](https://developer.apple.com/documentation/xctest/xctfail)
- [📒 Blog: XCTAssert Failure Messages](https://www.basbroek.nl/xctassert-asterisk)

serena0720 committed:
add: box office sample JSON 파일 추가

bubblecocoa committed:
feat: 폴더구조 변경 및 BoxOfficeResult 구조체 생성

serena0720 committed:
feat: DailyBoxOffice 구조체 생성

serena0720 committed:
refactor: [DailyBoxOffice] 타입 변경 및 private 접근 제어자 추가

bubblecocoa committed:
test: BoxOfficeTests, DailyBoxOfficeTests 파일 생성

serena0720 committed:
feat: JSONDecoder+ 생성

bubblecocoa committed:
feat: BoxOffice 구조체 생성

serena0720 committed:
refactor: BoxOfficeResult와 DailyBoxOffice 프로퍼티명 및 코딩키 수정

serena0720 committed:
test: BoxOfficeTests에 테스트케이스 작성

test_box_office_sample을_Decode하면_BoxOfficeResult_인스턴스가_존재한다

bubblecocoa committed:
test: coverage 확인을 위해 테스트플랜 추가

bubblecocoa committed:
chore: 작업자 이름 오탈자 수정

bubblecocoa committed:
test: BoxOfficeTests 테스트 변경

test_box_office_sample을_Decode하면_BoxOfficeResult_인스턴스가_존재한다() -> test_box_office_sample_json_파일을_디코딩_할_수_있다()

bubblecocoa committed:
test: BoxOfficeResultTests 생성 및 테스트 케이스 작성

serena0720 committed:
test: dailyBoxOfficeList 테스트 추가

test_dailyBoxOfficeList의_값이_있다

GREENOVER commented on BoxOffice/DTO/DailyBoxOffice.swift:
```diff
@@ -0,0 +1,48 @@
+//  Created by Hyun A Song on 2023/07/24.
+//
+
+struct DailyBoxOffice: Decodable {
```
나중에 저 필드들을 오랜만에 보면 실제 어떤 필드인지 헷갈리는 경우도 왕왕 있기에 그럴만한 필드들이 보인다면 설명 주석을 달아도 좋을듯 해보입니다 🙋🏻
예를들어, 제가 중간에 팀에 합류해서 아무것도 모르는 상태라 가정하고 (실제로도 그렇지만..!) movieNumber가 영화 갯수를 의미하는건지 영화의 번호를 의미하는건지 API 문서 등을 보지 않는다면 유추하기가 어려울것 같아서요 🥲
꼭 수정해야될 부분은 아니고 한번 요런 사람도 있겠구나~하고 생각해보면 좋을 포인트여서 남겨봐요!

bubblecocoa replied:
해당 필드에 대한 주석 남겼습니다!
의견 주신 부분에 격하게 공감합니다 🥹

GREENOVER commented on BoxOffice/DTO/BoxOfficeResult.swift:
```diff
@@ -0,0 +1,18 @@
+    let showRange: String
+    let dailyBoxOfficeList: [DailyBoxOffice]
+    
+    private enum CodingKeys: String, CodingKey {
```
프라이빗으로 정의하신 이유가 궁금합니다~ 🙋🏻
(잘못되었다는 소리 절대 아님 ❌)

serena0720 replied:
코딩키의 경우 내부적으로 설정하면 외부에서 접근을 할 필요가 없다고 생각하여 `private`을 붙여줬습니다!

GREENOVER commented on BoxOffice/Extension/JSONDecoder+.swift:
```diff
@@ -0,0 +1,20 @@
+
+extension JSONDecoder {
+    static func decode<T: Decodable>(fileName: String) -> T? {
+        let decoder = JSONDecoder()
+        
+        guard let asset = NSDataAsset(name: fileName) else { return nil }
+        
+        guard let data = try? decoder.decode(T.self, from: asset.data) else { return nil }
+        
+        return data
```
여러분들은 개행을 어떤 기준으로 해주시는지 궁금해요 🙋🏻

bubblecocoa replied:
1. `{}`가 생기는 경우 위아래로 빈줄을 삽입합니다.
2. `guard`문의 경우 `else` 뒤의 로직이나 제어구문이 짧은 경우 별도 개행을 하진 않고있습니다.
3. 메서드의 경우 `return`위에 로직이 존재하는 경우 `return`과 로직 사이에 개행을 넣습니다.
4. 로직이 길어질 경우 흐름이 바뀌는 경우 로직 사이에 빈줄을 삽입해주고 있습니다.

GREENOVER commented on BoxOffice/Extension/JSONDecoder+.swift:
```diff
@@ -0,0 +1,20 @@
+import UIKit
+
+extension JSONDecoder {
+    static func decode<T: Decodable>(fileName: String) -> T? {
```
디코딩을 실패하게 되면 에러 처리는 어떻게 생각하고 계시나요~?
Result 타입을 반환하여 해당 디코드 메서드에서 에러를 던져줘서 상위에서 처리할 수도 있을것 같기도 해보이는데 에러 처리에 대해 구상하신게 있으시면 공유 부탁드립니다 🙋🏻
만약 요 스텝에서 생각하지 않으셨던 포인트라면 요구사항은 아니지만 기초 작업에서부터 그런 포인트들은 생각해보셔도 좋을듯 합니다 🙌

bubblecocoa replied:
https://github.com/yagom-academy/ios-box-office/pull/72/commits/8b0b064fd1f4a0e0389ce4b933e22483e71de283, https://github.com/yagom-academy/ios-box-office/pull/72/commits/cf9ba742183bfb82041798f521468dda9e1ebe41
해당부분 수정했습니다!

GREENOVER commented on BoxOffice.xcodeproj/project.pbxproj:
```diff
@@ -49,22 +81,77 @@
+			isa = PBXGroup;
+			children = (
+				63DF20FC2970E1A1005DF7D1 /* Info.plist */,
+				CBC61F8B2A6EA40900FDAD18 /* App */,
+				CBC61F8C2A6EA41B00FDAD18 /* ViewController */,
+				CBC61F8D2A6EA42D00FDAD18 /* DTO */,
+				CBC61F8E2A6EA46400FDAD18 /* Resource */,
+				D2E159632A6F55E900052F8B /* Extension */,
```
크.. 그룹핑 아주 잘해주셨네요 👍👍👍

GREENOVER commented on BoxOfficeTests/BoxOfficeResultTests.swift:
```diff
@@ -0,0 +1,55 @@
+@testable import BoxOffice
+
+final class BoxOfficeResultTests: XCTestCase {
+    var sut: BoxOfficeResult!
```
테스트 코드긴 하지만, 강제 언래핑 하는 방식은 지양해보면 어떨까요~?

bubblecocoa replied:
해당 부분 `?`로 바꾸었습니다!

GREENOVER commented on BoxOfficeTests/BoxOfficeTests.swift:
```diff
@@ -0,0 +1,35 @@
+        try super.tearDownWithError()
+    }
+    
+    func test_box_office_sample_json_파일을_디코딩_할_수_있다() {
```
파일이 없어서 못하는 경우와 파일은 있는데 디코딩에 실패하는 케이스 둘 다 테스트 해볼 수 있을까요~? 🙋🏻

serena0720 replied:
https://github.com/yagom-academy/ios-box-office/pull/72/commits/4292cb1f3c5ccb6ea3b5e531a5bdd0bed8fe4ffc

`do-catch` 구문과` XCTAssertNotNil`과 `XCTFail`을 활용하여 수정하였습니다!

GREENOVER commented on BoxOfficeTests/BoxOfficeResultTests.swift:
```diff
@@ -0,0 +1,55 @@
+        XCTAssertEqual(result, expectedResult)
+    }
+    
+    func test_dailyBoxOfficeList의_값이_있다() {
```
추가로 해당하는 BoxOfficeResult 내 DailyBoxOffice들 중 정상적으로 데이터가 들어오고 현재 DailyBoxOffice와 매칭한 값과 맞는지 테스트하는것도 추가되면 더 완벽할것 같아요 🙋🏻

serena0720 replied:
https://github.com/yagom-academy/ios-box-office/pull/72/commits/4292cb1f3c5ccb6ea3b5e531a5bdd0bed8fe4ffc

`do-catch` 구문과` XCTAssertNotNil`과 `XCTFail`을 활용하여 수정하였습니다!

serena0720 replied:
꼼꼼한 코멘트 감사합니다! 해당 내용 테스트 생성하였습니다!

GREENOVER commented on BoxOffice/Extension/JSONDecoder+.swift:
```diff
@@ -0,0 +1,20 @@
+//
+//  JSONDecoder+.swift
```
오호 이와 같이 익스텐션 파일들에 대해 컨벤션을 가져가신 이유 혹은 어떤것들을 참고했는지 궁금합니다 🙋🏻

serena0720 replied:
이전 프로젝트들을 진행하면서 extension 파일을 해당 컨벤션으로 자주 사용하는 듯 하여 참고하였습니다!

GREENOVER reviewed (changes requested):
@serena0720 @bubblecocoa 
안녕하세요ㅎㅎ 
이번 박스오피스 프로젝트 같이 진행하게된 그린입니다 🙋🏻
다시 한번 저도 인사드리면서 2주간 잘 부탁드려요!!
박스 오피스 부셔봐요...😉
우선, 스텝 1 빠르고 또 요구사항에 맞게 잘 구현해주시느라 고생 많으셨습니다!
스텝 1은 요구사항이 크게 많은것이 없는데도 잘 놓치지 않고 해주셔서 리뷰를 크게 드릴건 없을것 같고
여러분들의 기준이나 의견이 궁금한 포인트들과 테스트 코드에서 좀 더 보완되면 좋을것 같은 부분에 대해 코멘트 남겼습니다 🙌
***
### 칭찬 드리고 싶은 부분
- 커밋 메시지 이력 관리를 너무 통일성 있고 의미있게 하셨네요 💯
- 파일 및 폴더 구조를 잘 잡고 그룹화도 너무 명확하게 되어있네요 💯
- DTO 객체 구분과 파일로도 명확하게 잘 분리시켜서 코드 파악이 너무 좋아요 💯
- 테스트 코드의 다양한 메서드들을 잘 활용하셨네요 💯
***
### 보완 혹은 생각해보면 좋을 포인트
- 테스트 코드 쪽에서 코멘트로 남기긴 했는데 거창하게 보완될 부분보다는 하나의 의견으로 보시고 여러분들의 기준과 의도에 맞게 한번 더 생각해보시면 아주 좋을것 같아요 👍
- 에러 처리에 대해서도 한번 고민해보셔도 좋을듯 합니다 🙋🏻
***
두분 고생 많으셨습니다!
잘 아시겠지만 제 리뷰는 정답보다는 같이 프로젝트를 진행하는 개발자의 하나의 의견으로 보시고 아무리 옳아보여도 의심해보시면서 여러분들의 기준과 근거에 맞는것만 채택해보면서 의미있는 코드를 만들어보시길 바랍니다!
요번 스텝에서 사실 요구사항을 다 만족하였기에 바로 어프로브 > 머지하여도 되지만 그래도 마지막 혹시 아쉬운게 있으실 수도 있어 점검차 Request Changes로 먼저 드리겠습니다 🥲
다음 스텝에서 생각해보고 리팩토링 하겠다 혹은 여기선 충분하다 하시면 언제든 머지 요청을 해주셔도 됩니다! ㅎㅎ
이번주도 힘내세요 💪

bubblecocoa committed:
comment: 필드에 필요한 설명 작성

serena0720 committed:
feat: JSONDecoderError 에러타입 추가 및 Test 적용

serena0720 committed:
Merge branch 'Step-1' of https://github.com/serena0720/ios-box-office into Step-1

bubblecocoa committed:
test: BoxOfficeResultTests의 setUpWithError() 내용 변경

serena0720 committed:
test: BoxOfficeResultTests 테스트 케이스 생성

test_box_office_sample과_일치하는_파일이_없어_Decode할_수_없다, test_box_office_sample를_이용하여_Decode가_실패한다

bubblecocoa committed:
test: BoxOfficeResultTests에 dailyBoxOffice 값에 대한 테스트 추가

test_dailyBoxOfficeList_첫번째_배열의_영화_제목이_경관의_피_이다
test_dailyBoxOfficeList_첫번째_배열의_영화_코드가_20199882_이다

GREENOVER reviewed (approved):
@serena0720 @bubblecocoa 
두분 리팩토링 하시느라 고생 많으셨어요!
아주 잘 보완되고 훨씬 더 좋아졌네요 👍
스텝 2로 넘어가시죠 ㅎㅎ
아..! 아실수도 있겠지만 참고로 리뷰 재요청을 해주실때 우측 상단 Reviewers 부분에서 리프레쉬를 눌러주시면 됩니다 🙌
![스크린샷 2023-07-26 오후 12 31 40](https://github.com/yagom-academy/ios-box-office/assets/72292617/b6c2852a-de46-4df8-9c7d-a471307f34dc)
스텝 2도 화이팅이에요~!
